.github/workflows/main.yml

```diff
@@ -0,0 +1,134 @@
+              root_changed="true"
+              echo "Root-level file changed: $file"
+            fi
+            if [[ "$file" =~ ^(python-flask/|c-sharp-razor/|rust-actix/|ruby-sinatra/|javascript-express/|go-gin/|go-gorilla/) ]]; then
+              any_service_changed="true"
+              echo "Service-level file changed: $file"
+            fi
+          done
+
+          # Set outputs
+          echo "::set-output name=root_changed::$root_changed"
+          echo "::set-output name=service_changed::$any_service_changed"
+
+  run-all-tests:
+    runs-on: ubuntu-latest
+    needs: check-root-changes
+    if: ${{ needs.check-root-changes.outputs.root_changed == 'true' }}
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - name: Set up environment and run all tests if root changed
+        run: |
+          echo "ROOT_CHANGED: ${{ needs.check-root-changes.outputs.root_changed }}"
+          echo "SERVICE_CHANGED: ${{ needs.check-root-changes.outputs.service_changed }}"
+          # Set up environment
+          sudo apt-get update
+          sudo apt-get install -y make python3-pip python3-venv
+          sudo curl -L "https://github.com/docker/compose/releases/download/v2.27.1/docker-compose-linux-x86_64" -o /usr/local/bin/docker-compose
+          sudo chmod +x /usr/local/bin/docker-compose
+
+          # Set up Python virtual environment
+          python3 -m venv venv
+          ./venv/bin/pip install --upgrade pip
+          ./venv/bin/pip install --no-cache-dir -r tests/requirements.txt
+
+          # Run all tests
+          source ./venv/bin/activate
+          make test-all
+
+  run-service-tests:
+    runs-on: ubuntu-latest
+    needs: check-root-changes
+    if: ${{ needs.check-root-changes.outputs.service_changed == 'true' && needs.check-root-changes.outputs.root_changed == 'false' }}
+    strategy:
+      matrix:
+        service:
+          - python-flask
+          - c-sharp-razor
+          - rust-actix
+          - ruby-sinatra
+          - javascript-express
+          - go-gin
+          - go-gorilla
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - name: Check if this service requires testing
+        id: service-check
+        run: |
+          if git diff --name-only origin/main...HEAD | grep -q "^${{ matrix.service }}/"; then
+            echo "::set-output name=service_changed::true"
+            echo "Service-specific changes detected for ${{ matrix.service }}"
+          else
+            echo "::set-output name=service_changed::false"
+            echo "No changes detected for ${{ matrix.service }}; skipping tests."
+          fi
+
+      - name: Set up environment and run tests if service changed
+        if: ${{ steps.service-check.outputs.service_changed == 'true' }}
+        run: |
+          # Set up environment
+          sudo apt-get update
```

Comment:
are you doing this whole process (setting the whole environment for each service)? Wouldn't be better to set the whole environment in the beginning and then just run for each service changed?

Reply:
Yeah I really tried doing that and spend hours on trying to store the environment in a cache but with no luck. So in the end I said fuck it since we just need some testing in the end.

Reply:
fair